src/openrct2-ui/windows/TrackDesignPlace.cpp

```diff
@@ -446,23 +500,19 @@ static int32_t window_track_place_get_base_z(int32_t x, int32_t y)
 static void window_track_place_attempt_placement(
     TrackDesign* td6, int32_t x, int32_t y, int32_t z, int32_t bl, money32* cost, ride_id_t* rideIndex)
 {
-    int32_t eax, ebx, ecx, edx, esi, edi, ebp;
-    money32 result;
-
-    edx = esi = ebp = 0;
-    eax = x;
-    ebx = bl;
-    ecx = y;
-    edi = z;
-
-    gActiveTrackDesign = _trackDesign.get();
-    result = game_do_command_p(GAME_COMMAND_PLACE_TRACK_DESIGN, &eax, &ebx, &ecx, &edx, &esi, &edi, &ebp);
-    gActiveTrackDesign = nullptr;
-
-    if (cost != nullptr)
-        *cost = result;
-    if (rideIndex != nullptr)
-        *rideIndex = edi & 0xFF;
+    auto tdAction = TrackDesignAction({ x, y, z, _currentTrackPieceDirection }, *_trackDesign);
+    tdAction.SetFlags(bl);
+    auto res = (bl & GAME_COMMAND_FLAG_APPLY) ? GameActions::Execute(&tdAction) : GameActions::Query(&tdAction);
+
+    if (res->Error != GA_ERROR::OK)
+    {
+        *cost = MONEY32_UNDEFINED;
+    }
+    else
+    {
+        *cost = res->Cost;
+    }
+    *rideIndex = dynamic_cast<TrackDesignActionResult*>(res.get())->rideIndex;
```

Comment:
This can crash if the `dynamic_cast` fails, wondering whether it should be protected.

Reply:
Checking for null can be avoided if the cast is moved into the branch where Error == GA_ERROR::OK and the other path should probably assign RIDE_ID_NULL. It also might be worth getting rid of this function and rather do this in-place.